qiskit_experiments/data_processing/data_processor.py

```diff
@@ -0,0 +1,118 @@
+    """
+
+    def __init__(self, input_key: str, data_actions: List[DataAction] = None):
+        """Create a chain of data processing actions.
+
+        Args:
+            input_key: The initial key in the datum Dict[str, Any] under which the data processor
+                will find the data to process.
+            data_actions: A list of data processing actions to construct this data processor with.
+                If None is given an empty DataProcessor will be created.
+        """
+        self._input_key = input_key
+        self._nodes = []
+
+        if data_actions:
+            for node in data_actions:
+                self.append(node)
+
+    def append(self, node: DataAction):
+        """
+        Append new data action node to this data processor.
+
+        Args:
+            node: A DataAction that will process the data.
+        """
+        self._nodes.append(node)
+
+    def __call__(self, datum: Dict[str, Any]) -> Tuple[Any, Any]:
+        """
+        Call self on the given datum. This method sequentially calls the stored data actions
+        on the datum.
+
+        Args:
+            datum: A single item of data, typically from an ExperimentData instance, that needs
+                to be processed. This dict also contains the metadata of each experiment.
+
+        Returns:
+            processed data: The data processed by the data processor.
+
+        Raises:
+            DataProcessorError: if no nodes are present.
+        """
+        if len(self._nodes) == 0:
+            raise DataProcessorError("Cannot call an empty data processor.")
+
+        datum_ = datum[self._input_key]
+
+        for node in self._nodes:
+            datum_ = node(datum_)
+
+        return datum_
+
+    def call_with_history(
+        self, datum: Dict[str, Any], history_nodes: Set = None
+    ) -> Tuple[Dict[str, Any], List]:
+        """
+        Call self on the given datum. This method sequentially calls the stored data actions
+        on the datum and also returns the history of the processed data.
+
+        Args:
+            datum: A single item of data, typically from an ExperimentData instance, that
+                needs to be processed.
+            history_nodes: The nodes, specified by index in the data processing chain, to
+                include in the history.
+
+        Returns:
+            processed data: The datum processed by the data processor.
+            history: The datum processed at each node of the data processor.
+
+        Raises:
+            DataProcessorError: If no nodes are present.
+        """
+        if len(self._nodes) == 0:
+            raise DataProcessorError("Cannot call an empty data processor.")
+
+        datum_ = datum[self._input_key]
+
+        history = []
+        for index, node in enumerate(self._nodes):
+            datum_ = node(datum_)
+
+            if history_nodes is None or (history_nodes and index in history_nodes):
```

Comment:
we can simply write
```python
history_nodes = history_nodes or set()
...
    if index in history_nodes:
        history.append((node.__class__.__name__, datum_, index))
```

Reply:
If we did this we lose the behavior in which giving None to `history_nodes` includes all nodes in the history. I would therefore not do this.

Reply:
Ah, okey. I understood what you intended. So `histroty_node=set()` and `history_node=None` behave differently. For me this looks complicated. Instead I would define `excluded_nodes` instead so that `None` and empty set will have the same meaning.

Reply:
The docstring clarifies this now. See here: cab9339